### PR TITLE
I2s experimental

### PR DESCRIPTION
**Pull request by Palatis**
various fixes to the i2s subsystem.
1. add missing convenient function (`i2s_write_lr_nb()`)
2. fix memory leak in `i2s_slc_begin()`
3. sample rate handling refactored

### CODE REVIEW

File: cores/esp8266/core_esp8266_i2s.c [hunk 1]
```diff
@@ -194,46 +219,94 @@ bool ICACHE_FLASH_ATTR i2s_write_lr(int16_t left, int16_t right){
   return i2s_write_sample(sample);
 }
 
+bool ICACHE_FLASH_ATTR i2s_write_lr_nb(int16_t left, int16_t right){
+  int sample = right & 0xFFFF;
```
Comment (me-no-dev) [inline cores/esp8266/core_esp8266_i2s.c]: why signed integer here and not uint32_t?

File: cores/esp8266/core_esp8266_i2s.c [hunk 2]
```diff
@@ -194,46 +219,94 @@ bool ICACHE_FLASH_ATTR i2s_write_lr(int16_t left, int16_t right){
...
   _i2s_sample_rate = rate;
-  uint32_t i2s_clock_div = (I2SBASEFREQ/(_i2s_sample_rate*32)) & I2SCDM;
-  uint8_t i2s_bck_div = (I2SBASEFREQ/(_i2s_sample_rate*i2s_clock_div*2)) & I2SBDM;
-  //os_printf("Rate %u Div %u Bck %u Frq %u\n", _i2s_sample_rate, i2s_clock_div, i2s_bck_div, I2SBASEFREQ/(i2s_clock_div*i2s_bck_div*2));
 
-  //!trans master, !bits mod, rece slave mod, rece msb shift, right first, msb right
-  I2SC &= ~(I2STSM | (I2SBMM << I2SBM) | (I2SBDM << I2SBD) | (I2SCDM << I2SCD));
-  I2SC |= I2SRF | I2SMR | I2SRSM | I2SRMS | ((i2s_bck_div-1) << I2SBD) | ((i2s_clock_div-1) << I2SCD);
+  // Find closest divider
+  uint32_t cpu_freq = ets_get_cpu_frequency() * 1000000L;
```
Comment (me-no-dev) [inline cores/esp8266/core_esp8266_i2s.c]: have you confirmed this to be related to the cpu clock?

File: cores/esp8266/i2s.h [hunk 3]
```diff
@@ -40,17 +40,22 @@ speed.
...
+void i2s_deinit();
 void i2s_begin();
 void i2s_end();
 void i2s_set_rate(uint32_t rate);//Sample Rate in Hz (ex 44100, 48000)
 bool i2s_write_sample(uint32_t sample);//32bit sample with channels being upper and lower 16 bits (blocking when DMA is full)
 bool i2s_write_sample_nb(uint32_t sample);//same as above but does not block when DMA is full and returns false instead
 bool i2s_write_lr(int16_t left, int16_t right);//combines both channels and calls i2s_write_sample with the result
+bool i2s_write_lr_nb(int16_t left, int16_t right);//same as above but does not block when DMA is full and returns false instead
+uint32_t * i2s_get_buffer(); // buffer available for writing, nullptr if not applicatable.
+void i2s_put_buffer(); // yup you guessed it.
```
Comment (me-no-dev) [inline cores/esp8266/i2s.h]: did not guess it ;)

File: cores/esp8266/core_esp8266_i2s.c [hunk 4]
```diff
@@ -194,46 +219,94 @@ bool ICACHE_FLASH_ATTR i2s_write_lr(int16_t left, int16_t right){
...
+
+  // CLK_I2S = CPU_FREQ / I2S_CLKM_DIV_NUM
+  // BCLK = CLK_I2S / I2S_BCK_DIV_NUM
+  // WS = BCLK/ 2 / (16 + I2S_BITS_MOD)
+  // Note that I2S_CLKM_DIV_NUM must be >5 for I2S data
+  // I2S_CLKM_DIV_NUM - 5 - 63
+  // I2S_BCK_DIV_NUM - 2 - 63
+  for (int bckdiv = 2; bckdiv < 64; bckdiv++) {
+    for (int clkmdiv = 5; clkmdiv < 64; clkmdiv++) {
+      uint32_t testfreq = cpu_freq / (bckdiv * clkmdiv * 32);
```
Comment (me-no-dev) [inline cores/esp8266/core_esp8266_i2s.c]: can't this be done with less iterations? like check the original calculation and the values around it?

**Comment by Palatis** [comment]
the clock divider guessing code is copied from mp3 example.
it use 160Mhz statically.

the modifications I made are reading runtime clock, and do not allow bit mode (like sending with 19 bit or 20 bit per sample, and let the DAC output at a lower volume.)

I'm still waiting for my hw college to provide me with a PCB with a working i2s DAC and amp, tho.


`i2s_write_lr()` takes two int16_t, so I think the non-blocking version should take signed as well.

**Comment by Palatis** [comment]
as for the i2s base clock, I confirmed that the samples are pushed out at almost correct rate, such as RATE_HZ samples pushed out around ~1015ms.

I think some rounding error and dma buffer consume speed and maybe some randomness are acceptable......

**Comment by me-no-dev** [comment]
I have also started writing this lib from the mp3 example ;) did many calculations and came up that the formula I used was the fastest and easiest option, but please prove me wrong :) we are people and make mistakes.
your left and right channels are signed 16 bit values, but both combined, it's unsigned 32 bit value ;)
do not remember if I run the tests at both 80 and 160 MHz, but I would think that I have and that is why I asked about the base clock in your calculations

**Comment by Palatis** [comment]
I will provide a easy sketch to test the sample rate sometime later.
Not at home right now.

File: cores/esp8266/i2s.h (hunk 3 above)
Comment (Palatis) [inline cores/esp8266/i2s.h]: sorry should provide better documentations...

the basic idea is call `i2s_get_buffer()`, and after you're done with the buffer, call `i2s_put_buffer()` to return the buffer to the subsystem.
maybe this should be called `i2s_commit_buffer()` instead.

current i2s has 8 of 64x4 bytes buffer, that's an undocumented secret to the user.
i'll improve the buffer initialization routines maybe lator.

my idea is to modify `i2s_init(size_t count, size_t size, size_t channels, size_t sizeof_sample)` to malloc() `count` of `size * channels * sizeof_sample` bytes buffers.

File: cores/esp8266/core_esp8266_i2s.c (hunk 4 above)
Comment (Palatis) [inline cores/esp8266/core_esp8266_i2s.c]: I ran a test program
```
#include <iostream>
#include <cstdint>

#define BASEFREQ (160L * 1000L * 1000L)

void i2s_set_rate(int rate, bool bits_mod) {
	//Find closest divider 
	int bestclkmdiv, bestbckdiv, bestbits, bestfreq=0;
	int tstfreq;
	/*
	   CLK_I2S = 160MHz / I2S_CLKM_DIV_NUM
	   BCLK = CLK_I2S / I2S_BCK_DIV_NUM
	   WS = BCLK/ 2 / (16 + I2S_BITS_MOD)
	   Note that I2S_CLKM_DIV_NUM must be >5 for I2S data
	   I2S_CLKM_DIV_NUM - 5-63
	   I2S_BCK_DIV_NUM - 2-63

	   We also have the option to send out more than 2x16 bit per sample. Most I2S codecs will
	   ignore the extra bits and in the case of the 'fake' PWM/delta-sigma outputs, they will just lower the output
	   voltage a bit, so we add them when it makes sense. Some of them, however, won't accept it, that's
	   why we have the option not to do this.
	 */
	for (int bckdiv=2; bckdiv < 64; bckdiv++) {
		for (int clkmdiv = 5; clkmdiv < 64; clkmdiv++) {
			for (int bits = 16; bits < (bits_mod ? 17 : 20); bits++) {
				tstfreq=BASEFREQ/(bckdiv*clkmdiv*bits*2);
				if (std::abs(rate-tstfreq)<std::abs(rate-bestfreq)) {
					bestfreq=tstfreq;
					bestclkmdiv=clkmdiv;
					bestbckdiv=bckdiv;
					bestbits=bits;
				}
			}
		}
	}

	std::cout
		<< "ReqRate " << rate << ", "
		<< "MclkDiv " << bestclkmdiv << ", "
		<< "BckDiv " << bestbckdiv << ", "
		<< "Bits " << bestbits << ", "
		<< "Freq " << (BASEFREQ / (bestbckdiv * bestclkmdiv * bestbits * 2)) << ", "
		<< "diff " << (rate - (BASEFREQ / (bestbckdiv * bestclkmdiv * bestbits * 2)))
		<< std::endl;
}

#define I2SCDM (0x3f)
#define I2SBDM (0x3f)
#define I2SBASEFREQ (12L * 1000L * 1000L)

void i2s_set_rate_static(int rate) {
	uint32_t i2s_clock_div = (I2SBASEFREQ/(rate*32)) & I2SCDM;
	uint32_t i2s_bck_div = (I2SBASEFREQ/(rate*i2s_clock_div*2)) & I2SBDM;

	std::cout
		<< "ReqRate " << rate << ", "
		<< "MclkDiv " << i2s_clock_div << ", "
		<< "BckDiv " << i2s_bck_div << ", "
		<< "Freq " << (I2SBASEFREQ / (i2s_clock_div * i2s_bck_div * 2)) << ", "
		<< "diff " << (rate - (I2SBASEFREQ / (i2s_clock_div * i2s_bck_div * 2)))
		<< std::endl;
}

int main() {
	int rates[] = { 8000, 11025, 16000, 22050, 32000, 44100, 48000, 88200, 96000 };
	for (int i = 0;i < sizeof(rates) / sizeof(*rates);++i) {
		std::cout << "mp3 demo (bits mod)   : ";
		i2s_set_rate(rates[i], true);
		std::cout << "mp3 demo (no bits mod): ";
		i2s_set_rate(rates[i], false);
		std::cout << "esp8266/Arduino       : ";
		i2s_set_rate_static(rates[i]);
		std::cout << std::endl;
	}
}
```
and got these result:
```
mp3 demo (bits mod)   : ReqRate 8000, MclkDiv 25, BckDiv 25, Bits 16, Freq 8000, diff 0
mp3 demo (no bits mod): ReqRate 8000, MclkDiv 25, BckDiv 25, Bits 16, Freq 8000, diff 0
esp8266/Arduino       : ReqRate 8000, MclkDiv 46, BckDiv 16, Freq 8152, diff -152

mp3 demo (bits mod)   : ReqRate 11025, MclkDiv 35, BckDiv 13, Bits 16, Freq 10989, diff 36
mp3 demo (no bits mod): ReqRate 11025, MclkDiv 31, BckDiv 13, Bits 18, Freq 11028, diff -3
esp8266/Arduino       : ReqRate 11025, MclkDiv 34, BckDiv 16, Freq 11029, diff -4

mp3 demo (bits mod)   : ReqRate 16000, MclkDiv 52, BckDiv 6, Bits 16, Freq 16025, diff -25
mp3 demo (no bits mod): ReqRate 16000, MclkDiv 49, BckDiv 6, Bits 17, Freq 16006, diff -6
esp8266/Arduino       : ReqRate 16000, MclkDiv 23, BckDiv 16, Freq 16304, diff -304

mp3 demo (bits mod)   : ReqRate 22050, MclkDiv 57, BckDiv 4, Bits 16, Freq 21929, diff 121
mp3 demo (no bits mod): ReqRate 22050, MclkDiv 38, BckDiv 5, Bits 19, Freq 22160, diff -110
esp8266/Arduino       : ReqRate 22050, MclkDiv 17, BckDiv 16, Freq 22058, diff -8

mp3 demo (bits mod)   : ReqRate 32000, MclkDiv 52, BckDiv 3, Bits 16, Freq 32051, diff -51
mp3 demo (no bits mod): ReqRate 32000, MclkDiv 49, BckDiv 3, Bits 17, Freq 32012, diff -12
esp8266/Arduino       : ReqRate 32000, MclkDiv 11, BckDiv 17, Freq 32085, diff -85

mp3 demo (bits mod)   : ReqRate 44100, MclkDiv 57, BckDiv 2, Bits 16, Freq 43859, diff 241
mp3 demo (no bits mod): ReqRate 44100, MclkDiv 19, BckDiv 5, Bits 19, Freq 44321, diff -221
esp8266/Arduino       : ReqRate 44100, MclkDiv 8, BckDiv 17, Freq 44117, diff -17

mp3 demo (bits mod)   : ReqRate 48000, MclkDiv 52, BckDiv 2, Bits 16, Freq 48076, diff -76
mp3 demo (no bits mod): ReqRate 48000, MclkDiv 49, BckDiv 2, Bits 17, Freq 48019, diff -19
esp8266/Arduino       : ReqRate 48000, MclkDiv 7, BckDiv 17, Freq 50420, diff -2420

mp3 demo (bits mod)   : ReqRate 88200, MclkDiv 19, BckDiv 3, Bits 16, Freq 87719, diff 481
mp3 demo (no bits mod): ReqRate 88200, MclkDiv 24, BckDiv 2, Bits 19, Freq 87719, diff 481
esp8266/Arduino       : ReqRate 88200, MclkDiv 4, BckDiv 17, Freq 88235, diff -35

mp3 demo (bits mod)   : ReqRate 96000, MclkDiv 26, BckDiv 2, Bits 16, Freq 96153, diff -153
mp3 demo (no bits mod): ReqRate 96000, MclkDiv 7, BckDiv 7, Bits 17, Freq 96038, diff -38
esp8266/Arduino       : ReqRate 96000, MclkDiv 3, BckDiv 20, Freq 100000, diff -4000
```

looks like the codes from mp3 demo does produce better `bck_div` and `clkm_div`.

Comment (Palatis) [inline cores/esp8266/core_esp8266_i2s.c]: maybe we just hard-code these with C++ template meta-programming, and only calculate those ones not hard-coded?

but anyway the loop can be finished within few ms, and it's not called really often.

File: cores/esp8266/core_esp8266_i2s.c (hunk 2 above)
Comment (Palatis) [inline cores/esp8266/core_esp8266_i2s.c]: sorry, I confirmed this **DOESN'T** relate to the cpu clock.

the codes with `ets_get_cpu_frequency()` runs exactly double the speed at 80Mhz than 160Mhz.

File: cores/esp8266/core_esp8266_i2s.c (hunk 1 above)
Comment (Palatis) [inline cores/esp8266/core_esp8266_i2s.c]: `i2s_write_lr()` does use int32_t, I copied the code and only changed that `i2s_write_sample(sample);` to `i2s_write_sample_nb(sample)`;

**Comment by earlephilhower** [comment]
Hi @Palatis I see this has been sitting in the backlog for two years, ouch.  Are you still interested in getting this patch into 2.5 or 2.6 of the Arduino SDK?  There was a massive cleanup this past release and we're trying to be more methodical in getting these PRs reviewed and accepted in a reasonable process and time.

If you're still interested, please update the conflicts and I'll do a review ASAP.

Thx
-EFP3

**Comment by Palatis** [comment]
i'm still willing to help, but i've shifted onto some other things since then.
I'm currently working in Shenzhen, but all my esp8266 (I'm referring to both hw and sw...) are at home in Taipei...

**Comment by earlephilhower** [comment]
No worries, don't want to put you out.  We can shelve this for now and if you ever want to get back to it we can re-open it.  Sound OK?

**Comment by Crypter** [comment]
I'll take a look in the following days, there sure must be something we can use and clean up this pull request too :)

**Comment by earlephilhower** [comment]
@Crypter and @Palatis, since the I2S subsystem has been rewritten to support bidirectional transfer and the method for choosing the proper clock setup for a given frequency has also been redone, I think this PR should be closed.

The PR is pretty much unmergeable now, and the only thing I'm seeing is missing is way of writing multiple samples-worth of data to the buffer (which is basically what the get/put buffer calls that were added let you do).  That's a much smaller PR and should be started against the current head to preserve everyone's sanity.  If I ever get time to get back to working on my ESP8266Audio library (which obviously uses the I2S extensively), adding that capability (uint32_t i2s_write_block/non(uint16_t samples[], int count));) is something I'll definitely look into as the overhead of calling the I2S add-sample call 44,000 times per second can't be small.

**Pull request closed without merging**